### PR TITLE
Fix Container.as_dict for case of flatten=True and add_prefix=True

### PR DESCRIPTION
**Pull request by maxnoe**
This change has the main goal of having
`ImageParametersContainer.as_dict(recursive=True, flatten=True,
add_prefix=True)` produce a dict with the same keys as the column names
in the dl1 parameters table in the hdf5 file.

Before:

```
{'params_hillas_hillas_intensity': nan,
 'params_hillas_hillas_skewness': nan,
 'params_hillas_hillas_kurtosis': nan,
 'params_hillas_hillas_fov_lon': <Quantity nan deg>,
 'params_hillas_hillas_fov_lat': <Quantity nan deg>,
 'params_hillas_hillas_r': <Quantity nan deg>,
 'params_hillas_hillas_phi': <Quantity nan deg>,
 'params_hillas_hillas_length': <Quantity nan deg>,
 'params_hillas_hillas_length_uncertainty': <Quantity nan deg>,
 'params_hillas_hillas_width': <Quantity nan deg>,
 'params_hillas_hillas_width_uncertainty': <Quantity nan deg>,
 'params_hillas_hillas_psi': <Quantity nan deg>,
 'params_timing_timing_intercept': nan,
 'params_timing_timing_deviation': nan,
 'params_timing_timing_slope': <Quantity nan 1 / deg>,
 'params_leakage_leakage_pixels_width_1': nan,
 'params_leakage_leakage_pixels_width_2': nan,
 'params_leakage_leakage_intensity_width_1': nan,
 'params_leakage_leakage_intensity_width_2': nan,
 'params_concentration_concentration_cog': nan,
 'params_concentration_concentration_core': nan,
 'params_concentration_concentration_pixel': nan,
 'params_morphology_morphology_num_pixels': -1,
 'params_morphology_morphology_num_islands': -1,
 'params_morphology_morphology_num_small_islands': -1,
 'params_morphology_morphology_num_medium_islands': -1,
 'params_morphology_morphology_num_large_islands': -1,
 'params_intensity_statistics_intensity_max': nan,
 'params_intensity_statistics_intensity_min': nan,
 'params_intensity_statistics_intensity_mean': nan,
 'params_intensity_statistics_intensity_std': nan,
 'params_intensity_statistics_intensity_skewness': nan,
 'params_intensity_statistics_intensity_kurtosis': nan,
 'params_peak_time_statistics_peak_time_max': nan,
 'params_peak_time_statistics_peak_time_min': nan,
 'params_peak_time_statistics_peak_time_mean': nan,
 'params_peak_time_statistics_peak_time_std': nan,
 'params_peak_time_statistics_peak_time_skewness': nan,
 'params_peak_time_statistics_peak_time_kurtosis': nan,
 'params_core_core_psi': <Quantity nan deg>}
```

After:

```
In [4]: c.as_dict(recursive=True, flatten=True, add_prefix=True)
Out[4]: 
{'hillas_intensity': nan,
 'hillas_skewness': nan,
 'hillas_kurtosis': nan,
 'hillas_fov_lon': <Quantity nan deg>,
 'hillas_fov_lat': <Quantity nan deg>,
 'hillas_r': <Quantity nan deg>,
 'hillas_phi': <Quantity nan deg>,
 'hillas_length': <Quantity nan deg>,
 'hillas_length_uncertainty': <Quantity nan deg>,
 'hillas_width': <Quantity nan deg>,
 'hillas_width_uncertainty': <Quantity nan deg>,
 'hillas_psi': <Quantity nan deg>,
 'timing_intercept': nan,
 'timing_deviation': nan,
 'timing_slope': <Quantity nan 1 / deg>,
 'leakage_pixels_width_1': nan,
 'leakage_pixels_width_2': nan,
 'leakage_intensity_width_1': nan,
 'leakage_intensity_width_2': nan,
 'concentration_cog': nan,
 'concentration_core': nan,
 'concentration_pixel': nan,
 'morphology_num_pixels': -1,
 'morphology_num_islands': -1,
 'morphology_num_small_islands': -1,
 'morphology_num_medium_islands': -1,
 'morphology_num_large_islands': -1,
 'intensity_max': nan,
 'intensity_min': nan,
 'intensity_mean': nan,
 'intensity_std': nan,
 'intensity_skewness': nan,
 'intensity_kurtosis': nan,
 'peak_time_max': nan,
 'peak_time_min': nan,
 'peak_time_mean': nan,
 'peak_time_std': nan,
 'peak_time_skewness': nan,
 'peak_time_kurtosis': nan,
 'core_psi': <Quantity nan deg>}
```

### CODE REVIEW

**Review by nbiederbeck — DISMISSED** [review]

File: ctapipe/core/container.py [hunk 1]
```diff
@@ -293,26 +293,27 @@ def as_dict(self, recursive=False, flatten=False, add_prefix=False):
...
-                        )
+
+        d = dict()
+        for key, val in self.items(add_prefix=add_prefix):
+            if isinstance(val, Container) or isinstance(val, Map):
+                if flatten:
+                    d.update(
+                        {
+                            k: v
+                            for k, v in val.as_dict(
```
Comment (nbiederbeck) [inline ctapipe/core/container.py:304]: codacy is right, you can just `d.update(val.as_dict(...` here

**Review by nbiederbeck — APPROVED** [review]

**Review by kosack — APPROVED** [review]

**Comment by LukasNickel** [comment]
Does this work with multiple algorithms predicting the same quantity (dl2, ml)?
Havent tested it yet, just my first random thought

**Comment by LukasNickel** [comment]
`is_valid` from the reconstructed shower container subcontainers for example

**Comment by maxnoe** [comment]
We should revisit this after #1949 is merged, changing to draft for now

**Comment by nbiederbeck** [comment]
The use-case for this change is still valid: testing with current master and this branch yields the same results as in the PR description. Therefore, I would suggest merging this, as there are no conflicts and already approved.

Additionally, this fixes a failing test I have on the `ml` branch.

**Pull request merged**